File: 34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.java
// Last updated: 8/13/2025, 8:24:08 PM
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int l = lowerBound(nums,target);
        int u = upperBound(nums,target);

        if(l==nums.length || nums[l]!=target){
            return new int[] {-1,-1};
        }
        else{
            return new int[] {l,u-1};
        }
    }

    public int upperBound(int []nums , int target){
        int low = 0 ;
        int high = nums.length-1;
        int ans = nums.length;
        while(low<=high){
            int mid = low + (high-low)/2;

            if(nums[mid]>target){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return ans;
    }

    public int lowerBound(int[] nums , int target){
        int low = 0;
        int high = nums.length -1;
        int ans = nums.length;
        while(low<=high){
            int mid = low + (high - low)/2;
            
            if(nums[mid]>=target){
                ans = mid;
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return ans;
    }
}

File: 34-FindFirstAndLastPositionOfElementInSortedArray/34-FindFirstAndLastPositionOfElementInSortedArray.py
# Last updated: 8/13/2025, 8:24:03 PM
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        first_position = self.first_position_finder(nums , target)
        second_position = self.second_position_finder(nums , target)
        return [first_position , second_position]

    
    def first_position_finder(self , nums : list[int] , target : int) -> int : 
        ans = -1
        low = 0
        high = len(nums)-1

        while low <= high : 
            mid = low + (high - low) // 2

            if nums[mid] == target : 
                ans = mid
                high = mid - 1
            
            elif nums[mid] > target : 
                high = mid - 1
            
            else : 
                low = mid + 1
        
        return ans
    
    def second_position_finder(self , nums : list[int] , target : int) -> int : 
        ans = -1
        low = 0
        high = len(nums)-1

        while low <= high : 
            mid = low + (high - low) // 2

            if nums[mid] == target : 
                ans = mid
                low = mid + 1
            
            elif nums[mid] < target : 
                low = mid + 1

            else : 
                high = mid - 1
        
        return ans